opm/parser/eclipse/EclipseState/Schedule/ActionX.hpp

```diff
@@ -49,21 +53,27 @@ namespace Opm {
 */
 
 class DeckKeyword;
+class ActionContext;
 
 class ActionX {
 public:
-    ActionX(const std::string& name, size_t max_run, double max_wait);
-    explicit ActionX(const DeckKeyword& kw);
+    ActionX(const std::string& name, size_t max_run, double max_wait, std::time_t start_time);
+    ActionX(const DeckKeyword& kw, std::time_t start_time);
+    ActionX(const DeckRecord& record, std::time_t start_time);
 
     void addKeyword(const DeckKeyword& kw);
+    bool ready(std::time_t sim_time) const;
+    bool eval(std::time_t sim_time, const ActionContext& context);
 
-    const std::string& name() const;
-private:
-    std::string m_name;
-    size_t max_run;
-    double max_wait;
+    const std::string name;
+    const size_t max_run;
+    const double min_wait;
+    const std::time_t start_time;
```

Comment:
I didn't pick up on this before (sorry), but those constant data members may come back to hurt us.  They prevent the compiler from generating assignment operators (i.e., `operator=(const& ActionX& rhs)` and `operator=(ActionX&& rhs)`) which, in turn, will limit the available operations on containers of `ActionX` (e.g., class `Opm::Actions`).

If you need these members to behave as `const` I'd probably recommend making them `private` and then providing `const` qualified accessor functions for them instead.

Reply:
Ahhhh - the boilerplate - you got to love it; this language has a certain feeling of quiltwork by commitee ..,

> They prevent the compiler from generating assignment operators (i.e., operator=(const& ActionX& rhs) and operator=(ActionX&& rhs)) 

I will not argue this - your knowledge of this far outweighs mine; but I do not understand it. The `operator= (const& ActionX& rhs)` should be perfectly safe to generate? I don't know the semantics of "moving out of"  a value type, but that should in my opinion be perfectly reducable to a no-op.

Anyway - I take your word at face value and will update with private members and so on.

Reply:
> > They prevent the compiler from generating assignment operators (i.e., operator=(const& ActionX& rhs) and operator=(ActionX&& rhs))

Doh, that should obviously be
```C++
operator=(const ActionX& rhs)
```
Sorry about the extra ampersand.

> I do not understand it. The `operator= (const ActionX& rhs)` should be perfectly safe to generate?

The default/compiler-generated assignment operator behaves like the following pseudo code
```C++
ActionX& ActionX::operator=(const ActionX& rhs)
{
    for (auto& member : this->members()) {
        this->member = rhs.member;
    }

    return *this;
}
```
If any of the members is `const`, then it cannot be assigned which means that the code above cannot do what it's supposed to do and is therefore not generated at all.  The default move-assignment operator looks the same except that the loop body is
```
this->member = std::move(rhs.member);
```
This matters if one of the members is a vector of databases, but falls back to simple assignment/copy if the member is just a built-in arithmetic type like an `int`.